### DISCUSSION TITLE
Multidex issue in new build process

### DISCUSSION THREAD

**Discussion by Kelsonk** [airsdk/Adobe-Runtime-Support#2101] [General]
I've been pulling my hair out trying to update everything to the new gradle based system. When I build my apk (aab will come later), I get this error in logcat.

08-26 16:06:40.727 32189 32189 E AndroidRuntime: Process: air.com.snowflake.CasteClicker, PID: 32189
08-26 16:06:40.727 32189 32189 E AndroidRuntime: java.lang.RuntimeException: Unable to instantiate application android.support.multidex.MultiDexApplication: java.lang.ClassNotFoundException: Didn't find class "android.support.multidex.MultiDexApplication" on path: DexPathList[[zip file "/data/app/air.com.snowflake.CasteClicker-U_jw36tzng-xtacAPDNYMA==/base.apk"],nativeLibraryDirectories=[/data/app/air.com.snowflake.CasteClicker-U_jw36tzng-xtacAPDNYMA==/lib/arm64, /data/app/air.com.snowflake.CasteClicker-U_jw36tzng-xtacAPDNYMA==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64]]
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.LoadedApk.makeApplication(LoadedApk.java:1230)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6551)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.ActivityThread.access$1400(ActivityThread.java:224)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1887)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:107)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:224)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7560)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
08-26 16:06:40.727 32189 32189 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "android.support.multidex.MultiDexApplication" on path: DexPathList[[zip file "/data/app/air.com.snowflake.CasteClicker-U_jw36tzng-xtacAPDNYMA==/base.apk"],nativeLibraryDirectories=[/data/app/air.com.snowflake.CasteClicker-U_jw36tzng-xtacAPDNYMA==/lib/arm64, /data/app/air.com.snowflake.CasteClicker-U_jw36tzng-xtacAPDNYMA==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64]]
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:230)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.AppComponentFactory.instantiateApplication(AppComponentFactory.java:76)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at androidx.core.app.CoreComponentFactory.instantiateApplication(CoreComponentFactory.java:52)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.Instrumentation.newApplication(Instrumentation.java:1156)
08-26 16:06:40.727 32189 32189 E AndroidRuntime:        at android.app.LoadedApk.makeApplication(LoadedApk.java:1222)

I've never had issues with multidex before and not sure how to go about fixing this. This shouldn't be an issue with any ANE right?

**Comment by Kelsonk** [comment]
Also I saw this thread but it doesn't really help me out, I'm running the newest version of the Adobe Air SDK

https://github.com/airsdk/Adobe-Runtime-Support/issues/102

**Reply by ajwfrost** [reply]
Hi

Yes that other thread was more about using the d8 tool rather than dx and those aren't used when the build is done with Gradle..

So the error you've got is because somewhere in your application you have a dependency on:
`android.support.multidex.MultiDexApplication`
We've moved over to the androidx classes now, so your ANE should instead refer to
`androidx.multidex.MultiDexApplication`

Do you have an ANE that overrides the application class? If so, this should be updated to use the androidx version; otherwise check your app descriptor file in case the application name is specified there.

If you really need to use the older version, then you'd need to build the `android-studio` output and then open this in Android Studio and update the app's `build.gradle` file to change the dependency on the androidx multidex library..

thanks

**Reply by Kelsonk** [reply]
Thanks a ton,  fixed that in my application xml. This is my new error, related to something about kotlin and the Facebook ane?

08-28 11:12:21.672 12732 12732 E AndroidRuntime: Process: air.com.snowflake.CasteClicker, PID: 12732
08-28 11:12:21.672 12732 12732 E AndroidRuntime: java.lang.NoClassDefFoundError: Failed resolution of: Lkotlin/jvm/internal/Intrinsics;
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at com.facebook.internal.FacebookInitProvider.<clinit>(FacebookInitProvider.kt:14)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at java.lang.Class.newInstance(Native Method)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.AppComponentFactory.instantiateProvider(AppComponentFactory.java:147)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at androidx.core.app.CoreComponentFactory.instantiateProvider(CoreComponentFactory.java:67)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.ActivityThread.installProvider(ActivityThread.java:7176)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6670)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6565)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.ActivityThread.access$1400(ActivityThread.java:224)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1887)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:107)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:224)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7560)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
08-28 11:12:21.672 12732 12732 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
08-28 11:12:21.672 12732 12732 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: kotlin.jvm.internal.Intrinsics

**Reply by Kelsonk** [reply]
I think I need to reread the docs and have some new things to implement for that ANE. Thanks for the help.

**Reply by Kelsonk** [reply]
I'm finally biting the bullet and moving everything to the apm system. It's pretty great! Thanks for the work you guys do.

Once question, I have multiple builds of very similar games that all use the same anes, the only difference is some of their properties. Is there a way to do different build configurations that not only changes the package name and version number, but also the properties of different anes? (the games have different facebook ids, ad identifier numbers, etc.)

**Reply by marchbold** [reply]
Great to hear `apm` is useful!  I use it for all our projects now and definitely has improved updates/installation of swc's and particular ane's for me. 

Regarding build configurations, they are currently designed to facilitate different configurations for the same application, eg debug/test/release types builds. You can override any of the parameters that are needed for extensions or swc's but not the core items currently, eg the application identifier or version number. However these could be good additions. Open an enhancement request on the apm issue tracker and I'll see if I can drop it into the next update. 

https://github.com/airsdk/apm/issues

**Reply by marchbold** [reply]
Just in case you hadnt' caught the existing functionality, it's documented here:

https://github.com/airsdk/apm/wiki/Usage-ProjectsAndPackages#build-types

**Comment by Kelsonk** [comment]
Some more info, this is just when trying to package a normal apk, when trying to package an aab I get this error after it builds and tries to install to the device.

Installing dist\ClickerNew.aab on the device...

file 'dist\ClickerNew.aab' does not contain AndroidManifest.xml
rm: /data/local/tmp/ClickerNew.aab: No such file or directory

**Reply by ajwfrost** [reply]
How are you trying to install it? and I'm slightly surprised you can create an AAB if you've got the earlier error .. but anyway: AAB files can't be directly installed onto a device, they need to be converted into APKS first. We do that automatically within ADT if you use the `-installApp` command.. otherwise you can use `bundletool` - or it looks like there are other utilities that do this too: https://beebom.com/how-install-android-app-bundles/

thanks
